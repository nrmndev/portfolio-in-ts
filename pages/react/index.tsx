import { Container, Row } from "react-bootstrap";
import { Routes, Route, NavLink } from "react-router-dom";
import TopicComponent from "./topic-component";
import { BiChevronRight } from "react-icons/bi";
import React, { useEffect } from "react";
import ReactIntroduction from "./introduction";

const DUMMY_LINKS = [
  {
    id: "1",
    title: "React Hooks",
    topics: [
      {
        topicTitle: "useState",
        path: "usestate",
        component: "usestate",
      },
      {
        topicTitle: "useEffect",
        path: "useeffect",
        component: "useeffect",
      },
      {
        topicTitle: "useRef",
        path: "useref",
        component: "useref",
      },
      {
        topicTitle: "useContext",
        path: "usecontext",
        component: "usecontext",
      },
      {
        topicTitle: "useCallback",
        path: "usecallback",
        component: "usecallback",
      },
      {
        topicTitle: "useMemo",
        path: "usememo",
        component: "usememo",
      },

      {
        topicTitle: "useImperativeHandle",
        path: "useimperativehandle",
        component: "useimperativehandle",
      },
      {
        topicTitle: "useReducer",
        path: "usereducer",
        component: "usereducer",
      },
      {
        topicTitle: "customhook",
        path: "customhook",
        component: "customhook",
      },
    ],
  },
  {
    id: "8",
    title: "React Router",
    topics: [
      {
        topicTitle: "Browser Router, Routes, Route",
        path: "react-router",
      },
      {
        topicTitle: "Route Guards",
        path: "react-createportal",
      },
    ],
  },

  {
    id: "2",
    title: "React Redux",
    topics: [
      {
        topicTitle: "Create Slice and Configure Store",
        path: "reduxcreateslice-configurestore",
      },
      { topicTitle: "Action Creators", path: "reduxactioncreators" },
      { topicTitle: "Typed Use Selectors", path: "reduxtypedusedselector" },
    ],
  },
  {
    id: "7",
    title: "Design Patterns",
    topics: [
      { topicTitle: "Higher Order Components", path: "designpattern-hoc" },
      {
        topicTitle: "Presentational vs Functional Component",
        path: "designpattern-presentational",
      },
      { topicTitle: "Provider Pattern", path: "designpattern-provider" },
      {
        topicTitle: "Conditional Rendering (Ternary Operator)",
        path: "designpattern-conditional",
      },
      {
        topicTitle: "Variant Components",
        path: "designpattern-variant",
      },
    ],
  },
  {
    id: "8",
    title: "Testing",
    topics: [
      { topicTitle: "Jest", path: "testing-jest" },
      {
        topicTitle: "React Testing Library",
        path: "testing-rtl",
      },
    ],
  },
  {
    id: "8",
    title: "Next JS",
    topics: [
      {
        topicTitle: "Server Side Rendering",
        path: "nextjs",
      },
    ],
  },

  {
    id: "6",
    title: "Code Base / Custom Utilities",
    topics: [
      { topicTitle: "Data Fetching", path: "codebase-fetching" },
      { topicTitle: "Data Pagination", path: "codebase-pagination" },
      { topicTitle: "Debouncing", path: "codebase-debouncing" },
      { topicTitle: "Authentication", path: "codebase-authentication" },
      { topicTitle: "Authorization", path: "codebase-authorization" },
      { topicTitle: "Form Validation", path: "codebase-formvalidation" },
      { topicTitle: "Create Portal", path: "codebase-createportal" },
      {
        topicTitle: "Local Storage and Session Storage",
        path: "codebase-storage",
      },
      {
        topicTitle: "Viewport (Responsive) Conditional Rendering",
        path: "codebase-viewport",
      },
    ],
  },

  {
    id: "9",
    title: "CSS Design Patterns",
    topics: [
      { topicTitle: "Module Based CSS", path: "css-modulebased" },
      {
        topicTitle: "Dynamic JS CSS",
        path: "css-dynamicjs",
      },
      {
        topicTitle: "Styled Components",
        path: "css-styledcomponents",
      },
      {
        topicTitle: "Theming SCSS/LESS",
        path: "css-theming-sass",
      },
      {
        topicTitle: "React Transition Group",
        path: "css-rtg",
      },
    ],
  },
  {
    id: "5",
    title: "Event Handlers",
    topics: [
      { topicTitle: "HTML Form Events", path: "eventhandler-form" },
      { topicTitle: "HTML Mouse Events", path: "eventhandler-mouse" },
    ],
  },
];

const ReactPage = () => {
  const links = DUMMY_LINKS.map((e, i) => {
    return (
      <li key={i}>
        <h3 className="h4 mt-2">{e.title}</h3>
        <ul>
          {e.topics.map((t, i) => {
            return (
              <li key={i}>
                <NavLink to={t.path}>
                  <BiChevronRight />
                  {t.topicTitle}
                </NavLink>
              </li>
            );
          })}
        </ul>
      </li>
    );
  });

  useEffect(() => {
    window.scrollTo(0, 0);
  });
  return (
    <>
      {
        <Container className="pt-5 pb-5">
          <h2 className="color_filled text__center font__h1 pt-5">
            ReactJS Skills Process
          </h2>
          <Row>
            <article className="col-xs-12 col-sm-8">
              <Routes>
                <Route path="/" element={<ReactIntroduction />} />
                <Route path=":filename" element={<TopicComponent />} />
              </Routes>
            </article>
            <aside className="col-xs-12 col-sm-4 scroll-y">
              <ul className="nav d-block">{links}</ul>
            </aside>
          </Row>
        </Container>
      }
    </>
  );
};

export default ReactPage;
