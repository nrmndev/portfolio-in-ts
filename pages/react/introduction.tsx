// import React from "react";
// import WithLineNumbers from "../../utils/code-highlighter";

// const exampleCode = `
// import React, { useState } from "react";

// const Example = () => {
//   const [count, setCount] = useState<number>(0);

//   const onClickHandler = (e: React.MouseEvent<HTMLButtonElement>): void => {
//     setCount(count+1);
//   }
//   return (
//     <div>
//       <p>You clicked {count} times</p>
//       <button onClick={() => setCount(count + 1)}>
//         Click me
//       </button>
//     </div>
//   );
// }
// `.trim();

const ReactIntroduction = () => {
  return (
    <>
      <h3>Getting Started</h3>
      <p>
        This page is an overview of all my skills process implemented using
        React and related resources.
      </p>
      <p>
        Feel free to take a look at other provided topics that you might find
        useful.
      </p>
    </>
  );
};

export default ReactIntroduction;
