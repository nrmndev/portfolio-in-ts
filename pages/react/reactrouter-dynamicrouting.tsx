//import React from "react";
//import WithLineNumbers from "../../utils/code-highlighter";

// const exampleCode = `
// import React, { useContext } from "react";
// import {NavLink} from "react-router-dom";
// import AuthContext from '../../store/auth/context';

// const AuthContext = React.createContext({
//   isLoggedIn: false
// })

// export default AuthContext
// `.trim();

const UseEffectSample = () => {
  return (
    <>
      <p className="color_1">Topic in progress</p>
    </>
  );
};

export default UseEffectSample;
