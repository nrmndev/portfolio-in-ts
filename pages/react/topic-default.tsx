const ReactDefault = () => {
  return <p>React default</p>;
};

export default ReactDefault;
