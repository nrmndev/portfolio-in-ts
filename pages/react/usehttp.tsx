import { useState, useCallback } from "react";

type requestConfig = {
  url: string;
  method?: string;
  headers?: { "Content-Type": string };
  body?: { text: string };
};
const useHttp = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const sendRequest = useCallback(
    async (requestConfig: requestConfig, applyData: any) => {
      setIsLoading(true);
      setError(null);
      try {
        const response = await fetch(requestConfig.url, {
          method: requestConfig.method ? requestConfig.method : "GET",
          headers: requestConfig.headers ? requestConfig.headers : {},
          body: requestConfig.body ? JSON.stringify(requestConfig.body) : null,
        });

        debugger;
        if (!response.ok) {
          throw new Error("Request failed!");
        }

        const data = await response.json();
        applyData(data);
      } catch (err) {
        if (err instanceof Error) {
          setError(err.message || "Something went wrong!");
        }
      }
      setIsLoading(false);
    },
    []
  );

  return {
    isLoading,
    error,
    sendRequest,
  };
};
export default useHttp;
