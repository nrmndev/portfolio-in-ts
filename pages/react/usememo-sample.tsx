import { useMemo } from "react";

// const useSortList:<Type> = (props: Type[]): Type[] => {
//   const sort: any = useMemo(
//     () => props.sort((a: Type[], b: Type) => a - b),
//     [props]
//   );
//   return sort.join();
// };

const useSortList = (props: number[]): number[] => {
  const sort: any = useMemo(
    () => props.sort((a: number, b: number) => a - b),
    [props]
  );
  return sort.join();
};
export default useSortList;
