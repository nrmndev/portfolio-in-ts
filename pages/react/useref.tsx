import { useEffect } from "react";
import WithLineNumbers from "../../utils/code-highlighter";
import { useRef } from "react";
const exampleCode = `
import React, { useRef } from "react";

const Example = () => {
  //Refs around HTML Elements, example for DIVs
  const divRef = useRef<HTMLDivElement>(null);
  divRef.current?.classList.add("REF");

  return (
    <>
      <div ref={divRef}>I have a classname called 'REF'</div>
    </>
  );
}
`.trim();

const exampleCode2 = `
import React, { useRef } from "react";

const Example = () => {
  const myFuncRef = useRef(() => {});

  myFuncRef.current = () => {
    console.log("I am logged");
  };

  useEffect(() => {
    //Fire the function
    myFuncRef.current();
  });
}
`.trim();

const UseEffectSample = () => {
  const myFuncRef = useRef(() => {});

  myFuncRef.current = () => {
    console.log("I am logged");
  };

  useEffect(() => {
    //Fire the function
    myFuncRef.current();
  });
  return (
    <>
      <div>Sample use of Ref on HTMLElements</div>
      <WithLineNumbers code={exampleCode} language="jsx" />
      <div>Sample use of Ref on Functions</div>
      <WithLineNumbers code={exampleCode2} language="jsx" />
    </>
  );
};

export default UseEffectSample;
