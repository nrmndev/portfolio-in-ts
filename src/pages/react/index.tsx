import { useCallback, useEffect, useState } from "react";
import { Routes, Route, NavLink } from "react-router-dom";
import { useAppDispatch } from "../../store/hooks/typedhooks";

import useHttp from "../../components/hooks/usehttp";
import { Container, Row } from "react-bootstrap";
import { BiChevronRight } from "react-icons/bi";
import ReactIntroduction from "./introduction";
import TopicComponent from "./topic-component";

type TopicsType = {
  id: string;
  title: string;
  topics: {
    topic_id: string;
    path: string;
    topicTitle: string;
  }[];
};

const ReactPage = () => {
  const { sendRequest } = useHttp();
  const [reactProcess, setReactProcess] = useState<TopicsType[]>([]);

  const setTopicsCallback = (data: TopicsType[]) => {
    setReactProcess(data);
  };

  const sendRequestCallback = useCallback(() => {
    const requestConfig = {
      url: `./react-process.json`,
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    };
    sendRequest(requestConfig, setTopicsCallback);
  }, [sendRequest]);

  const dispatch = useAppDispatch();
  useEffect(() => {
    //dispatch(fetchSkillsData()); //Activate Redux, currently using ContextAPI
    sendRequestCallback();
  }, [sendRequestCallback, dispatch]);

  const links =
    reactProcess &&
    reactProcess.length > 0 &&
    reactProcess.map((e: TopicsType) => {
      return (
        <li key={e.id}>
          <h3 className="h4 mt-2">{e.title}</h3>
          <ul>
            {e.topics &&
              e.topics.map((topicItem) => {
                return (
                  <li key={topicItem.topic_id}>
                    <NavLink to={topicItem.path}>
                      <BiChevronRight />
                      {topicItem.topicTitle}
                    </NavLink>
                  </li>
                );
              })}
          </ul>
        </li>
      );
    });

  useEffect(() => {
    window.scrollTo(0, 0);
  });

  // const validRouteLinks = reactProcess && reactProcess.length > 0 ? reactProcess
  //     .map((e) => e.topics)[0]
  //     .reduce((acc, curr) => {
  //       return acc.concat(curr.path);
  //     }, []) : "";
  //const validRouteLinks =: [""];
  //console.log(validRouteLinks);

  let validLinks;
  if (reactProcess && reactProcess.length > 0) {
    validLinks = reactProcess
      .map((e: TopicsType) => {
        return e.topics.reduce(
          (acc, curr) => acc.concat(curr.path),
          [] as string[]
        );
      })
      .flat(1);
  }
  console.log(validLinks);
  return (
    <>
      {
        <Container className="pt-5 pb-5">
          <h2 className="color_filled text__center font__h1 pt-5">
            ReactJS Skills Process
          </h2>
          <Row>
            <article className="col-xs-12 col-sm-8">
              <Routes>
                <Route path="/" element={<ReactIntroduction />} />
                <Route
                  path=":foldername/:filename"
                  element={<TopicComponent links={validLinks} />}
                />
              </Routes>
            </article>
            <aside className="col-xs-12 col-sm-4 scroll-y">
              <ul className="nav d-block">{links}</ul>
            </aside>
          </Row>
        </Container>
      }
    </>
  );
};

export default ReactPage;
