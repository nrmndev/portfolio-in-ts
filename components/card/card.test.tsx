import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import Card, { CARD_VARIANTS } from "./card.component";

describe("Card Component", () => {
  test("snapshot", () => {
    const card = render(
      <Card variant={CARD_VARIANTS.WITHSHADOW}>Some items</Card>
    );
    expect(card).toMatchSnapshot();
  });
  test("renders card StyledCardBase", () => {
    const tree = renderer.create(<Card>Some item</Card>).toJSON();
    expect(tree).toHaveStyleRule(`border-radius: 10px;
        padding: 35px 50px 35px 50px;
        background: linear-gradient(165deg, #fff 50%, #edebeb 80%);
        position: relative;
        overflow: hidden;`);
  });

  test("renders card StyledCardWithShadow", () => {
    //render(<Card>Some item</Card>);
    const tree = renderer
      .create(<Card variant={CARD_VARIANTS.WITHSHADOW}>Some item</Card>)
      .toJSON();
    expect(tree).toHaveStyleRule(`box-shadow: 2px 2px 15px #d8d5d5;`);
  });
});
