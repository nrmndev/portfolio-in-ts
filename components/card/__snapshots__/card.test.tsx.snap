// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c3 {
  -webkit-transition: -webkit-transform 0.3s ease-in;
  -webkit-transition: transform 0.3s ease-in;
  transition: transform 0.3s ease-in;
  -webkit-transform: translateY(10px);
  -ms-transform: translateY(10px);
  transform: translateY(10px);
}

.c0 {
  border-radius: 10px;
  padding: 35px 50px 35px 50px;
  background: linear-gradient(165deg,#fff 50%,#edebeb 80%);
  position: relative;
  overflow: hidden;
}

.c0:before {
  content: "";
  position: absolute;
  -webkit-transition: opacity 0.3s ease-in;
  transition: opacity 0.3s ease-in;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  top: 0;
  left: 0;
  background: linear-gradient(165deg,#fff 0%,#fafafa 50%);
  opacity: 0;
  z-index: 0;
}

.c0:hover .c2 {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c1 {
  box-shadow: 2px 2px 15px #d8d5d5;
}

<body>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c2 c3"
        >
          Some items
        </div>
      </div>
    </div>
  </body>,
  "container": .c3 {
  -webkit-transition: -webkit-transform 0.3s ease-in;
  -webkit-transition: transform 0.3s ease-in;
  transition: transform 0.3s ease-in;
  -webkit-transform: translateY(10px);
  -ms-transform: translateY(10px);
  transform: translateY(10px);
}

.c0 {
  border-radius: 10px;
  padding: 35px 50px 35px 50px;
  background: linear-gradient(165deg,#fff 50%,#edebeb 80%);
  position: relative;
  overflow: hidden;
}

.c0:before {
  content: "";
  position: absolute;
  -webkit-transition: opacity 0.3s ease-in;
  transition: opacity 0.3s ease-in;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  top: 0;
  left: 0;
  background: linear-gradient(165deg,#fff 0%,#fafafa 50%);
  opacity: 0;
  z-index: 0;
}

.c0:hover .c2 {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c1 {
  box-shadow: 2px 2px 15px #d8d5d5;
}

<div>
    <div
      class="c0 c1"
    >
      <div
        class="c2 c3"
      >
        Some items
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
